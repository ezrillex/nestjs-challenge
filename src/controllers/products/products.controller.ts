import { Controller, Get } from '@nestjs/common';
import { PublicPrivate } from '../../decorators/public_and_private/public_and_private.decorator';

@Controller('products')
export class ProductsController {
  @Get()
  @PublicPrivate()
  async getProducts() {
    return 'hello1!!!';
  }
}
