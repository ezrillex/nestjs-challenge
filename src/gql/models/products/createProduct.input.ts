import {
  ArrayMinSize,
  IsArray,
  IsString,
  ValidateNested,
} from 'class-validator';
import { Type } from 'class-transformer';
import { CreateVariationInput } from './createVariation.input';
import { Field, InputType } from '@nestjs/graphql';

@InputType()
export class CreateProductInput {
  @IsString()
  @Field()
  name: string;

  @IsString()
  @Field()
  description: string;

  @Field(() => [CreateVariationInput])
  @IsArray()
  @ValidateNested({ each: true })
  @ArrayMinSize(1)
  @Type(() => CreateVariationInput)
  variations: CreateVariationInput[];
}
