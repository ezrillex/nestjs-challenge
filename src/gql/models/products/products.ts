import { Field, ObjectType } from '@nestjs/graphql';
import { IsUUID } from 'class-validator';
import { ProductVariations } from './product-variations/product-variations';

@ObjectType()
export class Products {
  @Field(() => String)
  id: string;

  @Field(() => Boolean)
  is_published: boolean;

  @Field(() => Boolean)
  is_deleted: boolean;

  @Field(() => String)
  name: string;

  @Field(() => String)
  description: string;

  @Field(() => [ProductVariations])
  variations: ProductVariations[];

  @Field(() => String)
  created_by: string;

  @Field(() => Date)
  created_at: string;

  @IsUUID()
  @Field(() => String, { nullable: true })
  last_updated_by: string;

  @Field(() => String, { nullable: true })
  last_updated_at: string;
}
