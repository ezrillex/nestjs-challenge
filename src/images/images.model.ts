import { Field, ObjectType } from '@nestjs/graphql';

@ObjectType()
export class Images {
  @Field(() => String)
  id: string;
  @Field(() => String)
  url: string;

  @Field(() => Date)
  created_at: Date;

  @Field(() => String, { nullable: true })
  product_variation_id: string;
}
