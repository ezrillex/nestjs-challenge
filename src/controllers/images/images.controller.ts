import {
  BadRequestException,
  Body,
  Controller,
  Delete,
  Get,
  HttpCode,
  HttpStatus,
  NotFoundException,
  Param,
  ParseUUIDPipe,
  Post,
} from '@nestjs/common';
import { RequiresRole } from '../../decorators/requires-role/requires-role.decorator';
import { roles } from '@prisma/client';
import { CreateImageDto } from '../../DTOs/create-image-dto/create-image-dto';
import { ImagesService } from '../../services/images/images.service';
import { ProductsService } from '../../services/products/products.service';

@Controller('images')
export class ImagesController {
  constructor(
    private readonly imageService: ImagesService,
    private readonly productService: ProductsService,
  ) {}

  @RequiresRole(roles.manager)
  @Post()
  async createImage(@Body() data: CreateImageDto) {
    if (!data.image_mime_type.startsWith('image/')) {
      throw new BadRequestException('Image must have a valid image mime type');
    }

    const record = await this.productService.GetProductVariationById(
      data.product_variation_id,
    );

    if (!record) {
      throw new NotFoundException('Product Variation not found');
    }

    return this.imageService.uploadImage(
      data.image,
      data.image_mime_type,
      data.product_variation_id,
    );
  }

  @HttpCode(HttpStatus.NO_CONTENT)
  @RequiresRole(roles.manager)
  @Delete(':id')
  async deleteImage(@Param('id', ParseUUIDPipe) id: string) {
    // No return due to delete being no content by design.
    await this.imageService.deleteImageById(id);
  }
}
