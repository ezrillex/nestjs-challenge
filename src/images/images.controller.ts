import {
  Body,
  Controller,
  Delete,
  HttpCode,
  HttpStatus,
  Param,
  ParseUUIDPipe,
  Post,
} from '@nestjs/common';
import { RequiresRole } from '../common/decorators/requires-role.decorator';
import { roles } from '@prisma/client';
import { CreateImageDto } from './create-image.dto';
import { ImagesService } from './images.service';
import { Images } from './images.model';

@Controller('images')
export class ImagesController {
  constructor(private readonly imageService: ImagesService) {}

  @RequiresRole(roles.manager)
  @Post()
  async createImage(@Body() data: CreateImageDto): Promise<Images> {
    return this.imageService.uploadImage(
      data.image,
      data.image_mime_type,
      data.product_variation_id,
    );
  }

  @HttpCode(HttpStatus.NO_CONTENT)
  @RequiresRole(roles.manager)
  @Delete(':id')
  async deleteImage(@Param('id', ParseUUIDPipe) id: string): Promise<void> {
    await this.imageService.deleteImageById(id);
  }
}
