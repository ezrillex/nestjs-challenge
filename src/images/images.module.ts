import { Module } from '@nestjs/common';
import { ImagesController } from './images.controller';
import { ImagesService } from './images.service';
import { PrismaService } from '../prisma/prisma.service';
import { ProductsService } from '../products/products.service';

@Module({
  controllers: [ImagesController],
  providers: [ImagesService, PrismaService, ProductsService],
})
export class ImagesModule {}
