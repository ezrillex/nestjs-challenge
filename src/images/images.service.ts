import {
  BadRequestException,
  Injectable,
  InternalServerErrorException,
  NotFoundException,
} from '@nestjs/common';
import { v2 as cloudinary } from 'cloudinary';
import { ConfigService } from '@nestjs/config';
import { PrismaService } from '../prisma/prisma.service';
import { ProductsService } from '../products/products.service';
import { Images } from './images.model';

@Injectable()
export class ImagesService {
  constructor(
    private readonly configService: ConfigService,
    private readonly prisma: PrismaService,
    private readonly productsService: ProductsService,
  ) {
    cloudinary.config({
      cloud_name: configService.get<string>('CLOUDINARY_CLOUD_NAME'),
      api_key: configService.get<string>('CLOUDINARY_API_KEY'),
      api_secret: configService.get<string>('CLOUDINARY_API_SECRET'),
    });
  }

  async uploadImage(
    imageBase64: string,
    mime_type: string,
    product_variation_id: string,
  ): Promise<Images> {
    if (!mime_type.startsWith('image/')) {
      throw new BadRequestException('Image must have a valid image mime type');
    }

    const record = await this.productsService.getProductVariationById(
      product_variation_id,
      true,
    );
    if (record === 0) {
      throw new NotFoundException('Product Variation not found');
    }

    const image = await cloudinary.uploader.upload(
      `data:${mime_type};base64,${imageBase64}`,
    );
    if (!image.public_id) {
      throw new InternalServerErrorException(
        'There was an issue with the CDN, please try again later.',
      );
    }

    return this.prisma.images.create({
      select: {
        id: true,
        url: true,
        created_at: true,
        product_variation_id: true,
      },
      data: {
        cdn_id: image.public_id,
        url: image.secure_url,
        created_at: new Date().toISOString(),
        product_variation: {
          connect: { id: product_variation_id },
        },
      },
    });
  }

  async deleteImageById(id: string): Promise<Images> {
    const record = await this.prisma.images.findUnique({
      where: { id: id },
    });
    if (!record) {
      throw new NotFoundException('Image not found');
    }

    const cdn_response = await cloudinary.uploader.destroy(record.cdn_id);

    if (cdn_response && cdn_response.result && cdn_response.result === 'ok') {
      return this.prisma.images.delete({
        where: { id: record.id },
      });
    } else {
      throw new InternalServerErrorException(
        `An error occurred when deleting the image.`,
      );
    }
  }

  async getImagesByProductVariation(
    product_variation_id: string,
  ): Promise<Images[]> {
    const { images } = await this.prisma.productVariations.findUnique({
      where: { id: product_variation_id },
      select: {
        images: true,
      },
    });
    return images;
  }
}
