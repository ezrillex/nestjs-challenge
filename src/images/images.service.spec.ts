import { Test, TestingModule } from '@nestjs/testing';
import { ImagesService } from './images.service';
import { ConfigService } from '@nestjs/config';
import { PrismaService } from '../prisma/prisma.service';
import { ProductsService } from '../products/products.service';
import { v2 as cloudinary, UploadApiResponse } from 'cloudinary';

describe('ImagesService', () => {
  let service: ImagesService;
  let prismaService: PrismaService;
  let configService: ConfigService;
  let productService: ProductsService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [ConfigService, ProductsService, PrismaService, ImagesService],
    }).compile();

    service = module.get(ImagesService);
    configService = module.get(ConfigService);
    prismaService = module.get(PrismaService);
    productService = module.get(ProductsService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('config service should be defined', () => {
    expect(configService).toBeDefined();
  });

  it('prisma service should be defined', () => {
    expect(prismaService).toBeDefined();
  });

  it('prisma service should be defined', () => {
    expect(productService).toBeDefined();
  });

  describe('upload image method', () => {
    it('should error if not image mime type', async () => {
      await expect(
        service.uploadImage(
          'the_base_64',
          'video/webp',
          '2730fc05-6f87-49e5-8a41-559208048ebe',
        ),
      ).rejects.toThrowErrorMatchingSnapshot('image must be valid mime type');
    });

    it('should accept image mime type', async () => {
      await expect(
        service.uploadImage(
          'the_base_64',
          'image/jpeg',
          '2730fc05-6f87-49e5-8a41-559208048ebe',
        ),
      ).rejects.toThrowErrorMatchingSnapshot('product variation not found');
    });

    it('should error if invalid product id', async () => {
      await expect(
        service.uploadImage(
          'the_base_64',
          'image/jpeg',
          '2730fc05-6f87-49e5-8a41-559208048ebe',
        ),
      ).rejects.toThrowErrorMatchingSnapshot('product variation not found');
    });

    it('should error if cdn didnt provide a public id for the uploaded image. ', async () => {
      const mock_cdn_response = {
        public_id: undefined,
      } as UploadApiResponse;
      jest
        .spyOn(cloudinary.uploader, 'upload')
        .mockResolvedValue(mock_cdn_response);

      jest
        .spyOn(productService, 'GetProductVariationById')
        .mockResolvedValue(1);

      await expect(
        service.uploadImage(
          'the_base_64',
          'image/jpeg',
          '2730fc05-6f87-49e5-8a41-559208048ebe',
        ),
      ).rejects.toThrowErrorMatchingSnapshot(
        'CDN didnt provide the required id',
      );
    });
  });

  describe('delete image method', () => {
    it('should error if invalid image id', async () => {
      await expect(
        service.deleteImageById('2730fc05-6f87-49e5-8a41-559208048ebe'),
      ).rejects.toThrowErrorMatchingSnapshot('image not found');
    });

    it('should error if cdn failed and response was not successful ', async () => {
      const mock_cdn_response = {
        result: 'definitely not ok',
      };
      jest
        .spyOn(cloudinary.uploader, 'destroy')
        .mockResolvedValue(mock_cdn_response);

      const mock_record = {
        id: 'string',
        cdn_id: 'string',
        url: 'string',
        created_at: new Date(),
        product_variation_id: null,
      };
      jest
        .spyOn(prismaService.images, 'findUnique')
        .mockResolvedValue(mock_record);

      await expect(
        service.deleteImageById('2730fc05-6f87-49e5-8a41-559208048ebe'),
      ).rejects.toThrowErrorMatchingSnapshot('CDN result was not successful');
    });
  });
});
