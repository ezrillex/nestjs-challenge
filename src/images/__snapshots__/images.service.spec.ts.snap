// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImagesService deleteImageById should return successfully deleted image when CDN response is ok: request sent to prisma 1`] = `
[
  [
    {
      "where": {
        "id": "string",
      },
    },
  ],
]
`;

exports[`ImagesService deleteImageById should throw an error for invalid image id: image not found 1`] = `"Image not found"`;

exports[`ImagesService deleteImageById should throw an error when CDN fails and returns unsuccessful response: CDN result was not successful 1`] = `"An error occurred when deleting the image."`;

exports[`ImagesService uploadImage should create a record in the database after successful upload: query all ok 1`] = `
[
  [
    {
      "data": {
        "cdn_id": "8b3ae683-0626-44be-b591-9271e288388f",
        "created_at": "2021-01-12T18:12:00.000Z",
        "product_variation": {
          "connect": {
            "id": "2730fc05-6f87-49e5-8a41-559208048ebe",
          },
        },
        "url": "https://www.cloudinary.com/",
      },
      "select": {
        "created_at": true,
        "id": true,
        "product_variation_id": true,
        "url": true,
      },
    },
  ],
]
`;

exports[`ImagesService uploadImage should successfully accept valid image mime type: product variation not found 1`] = `"Product Variation not found"`;

exports[`ImagesService uploadImage should throw an error for invalid image mime type: image must be valid mime type 1`] = `"Image must have a valid image mime type"`;

exports[`ImagesService uploadImage should throw an error if CDN didnâ€™t provide the required public id: CDN didnt provide the required id 1`] = `"There was an issue with the CDN, please try again later."`;

exports[`ImagesService uploadImage should throw an error if product variation id is invalid: product variation not found 1`] = `"Product Variation not found"`;
